File: JavaScript/src/Internals/Functions/MathBase/Function_FIXED.js
import { Function_3 } from '../Function_3.js';
import { Operand } from '../../../Operand.js';

// 模拟 C# Math.Round(x, 0, MidpointRounding.AwayFromZero)
function roundAwayFromZero(x) {
    return x < 0 ? -Math.round(-x) : Math.round(x);
}

class Function_FIXED extends Function_3 {
    get Name() {
        return "Fixed";
    }

    constructor(z) {
        super(z);
    }

    evaluate(engine, tempParameter) {
        let args1 = this.getNumber_1(engine, tempParameter);
        if (args1.IsError) { return args1; }

        let num = 2;
        if (this.b !== null && this.b !== undefined) {
            let args2 = this.getNumber_2(engine, tempParameter);
            if (args2.IsError) { return args2; }
            num = args2.IntValue;
            // Excel 支持负数 decimals(向左取整),如 FIXED(1234.567,-1)="1,230",范围与 ROUND 一致
            if (num < -15 || num > 15) {
                return this.parameterError(2);
            }
        }

        let s = args1.NumberValue;
        if (num >= 0) {
            let factor = Math.pow(10, num);
            s = roundAwayFromZero(s * factor) / factor;
        } else {
            // Math.Round(decimal, int) 只支持非负位数,负数位数(向左取整)改用先除后乘
            let factor = Math.pow(10, -num);
            s = roundAwayFromZero(s / factor) * factor;
        }
        let no = false;
        if (this.c !== null && this.c !== undefined) {
            let args3 = this.getBoolean_3(engine, tempParameter);
            if (args3.IsError) { return args3; }
            no = args3.BooleanValue;
        }
        if (no === false) {
            // 负数位数取整后无小数位,用 0 位小数保持千分位
            let formatted = (num < 0 ? s.toFixed(0) : s.toFixed(num));
            let parts = formatted.split('.');
            parts[0] = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, ',');
            return Operand.Create(parts.join('.'));
        }
        return Operand.Create(s.toString());
    }
}

export { Function_FIXED };

File: java/src/main/java/toolgood/algorithm/internals/functions/mathbase/Function_FIXED.java
package toolgood.algorithm.internals.functions.mathbase;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.text.DecimalFormat;
import java.text.DecimalFormatSymbols;
import java.util.List;
import java.util.Locale;
import java.util.function.BiFunction;

import toolgood.algorithm.AlgorithmEngine;
import toolgood.algorithm.Operand;
import toolgood.algorithm.enums.OperandType;
import toolgood.algorithm.internals.ParameterType;
import toolgood.algorithm.internals.functions.FunctionBase;
import toolgood.algorithm.internals.functions.Function_3;
import toolgood.algorithm.internals.functions.NoneEngine;

public final class Function_FIXED extends Function_3 {
    public Function_FIXED(FunctionBase[] funcs) {
        super(funcs);
    }

    @Override
    public String Name() {
        return "Fixed";
    }

    @Override
    public Operand Evaluate(AlgorithmEngine engine, BiFunction<AlgorithmEngine, String, Operand> tempParameter) throws Exception {
        int num = 2;
        if (func2 != null) {
            Operand args2 = GetNumber_2(engine, tempParameter);
            if (args2.IsErrorOrNone()) {
                return args2;
            }
            num = args2.IntValue();
        }

        boolean noCommas = false;
        if (func3 != null) {
            Operand args3 = GetBoolean_3(engine, tempParameter);
            if (args3.IsErrorOrNone()) {
                return args3;
            }
            noCommas = args3.BooleanValue();
        }

        Operand args1 = GetNumber_1(engine, tempParameter);
        if (args1.IsErrorOrNone()) {
            return args1;
        }

        BigDecimal value = args1.NumberValue();
        value = value.setScale(num, RoundingMode.HALF_UP);

        if (noCommas) {
            return Operand.Create(value.toPlainString());
        }

        DecimalFormatSymbols symbols = new DecimalFormatSymbols(Locale.US);
        DecimalFormat df = new DecimalFormat("#,##0." + (num > 0 ? new String(new char[num]).replace('\0', '0') : ""), symbols);
        return Operand.Create(df.format(value));
    }

    @Override
    public OperandType GetResultType() {
        return OperandType.TEXT;
    }

    @Override
    public void GetParameterTypes(NoneEngine noneEngine, List<ParameterType> result, OperandType operandType,
            String op, String val) {
        func1.GetParameterTypes(noneEngine, result, OperandType.NUMBER);
        if (func2 != null) {
            func2.GetParameterTypes(noneEngine, result, OperandType.NUMBER);
        }
        if (func3 != null) {
            func3.GetParameterTypes(noneEngine, result, OperandType.BOOLEAN);
        }
    }
}
